db/util.go

```diff
@@ -35,11 +35,26 @@ func cpIncr(bz []byte) (ret []byte) {
 			ret[i] = byte(0x00)
 		}
 	}
-	return EndingKey()
+	return nil
 }
 
-func IsKeyInDomain(key, start, end []byte) bool {
-	leftCondition := bytes.Equal(start, BeginningKey()) || bytes.Compare(key, start) >= 0
-	rightCondition := bytes.Equal(end, EndingKey()) || bytes.Compare(key, end) < 0
-	return leftCondition && rightCondition
+// See DB interface documentation for more information.
+func IsKeyInDomain(key, start, end []byte, isReverse bool) bool {
+	if !isReverse {
+		if bytes.Compare(key, start) < 0 {
+			return false
+		}
+		if end != nil && bytes.Compare(end, key) <= 0 {
+			return false
+		}
+		return true
+	} else {
+		if start != nil && bytes.Compare(start, key) < 0 {
+			return false
+		}
+		if end != nil && bytes.Compare(key, end) <= 0 {
```

Comment:
Won't this condition always return false say
given:
key:   []byte("c")
start: []byte("a")
end:   []byte("z")
reverse: false

Given the docs at https://github.com/tendermint/tmlibs/blob/091b416370d448f03390c68dc2f663614cbaac36/db/types.go#L23-L28

Reply:
Or perhaps could you explain to me what IsKeyInDomain is supposed to do?

Reply:
if `isReverse==false` we're in the first branch,  so in the example you've given we'd return `true`